src/functions.php

```diff
@@ -230,6 +231,15 @@ function ($reason, $idx, Promise $aggregate) {
     )->then(function () use (&$results) {
         ksort($results);
         return $results;
+    })->then(function ($results) use ($recursive, &$promises) {
```

Comment:
Can you update the PR so that this extra `then` is only be added if `$recursive` is set to true? That would make this addition have almost zero cost over the current function for the default use case, saving the added overhead of the new promise and function.